ouroboros-network/src/Data/FingerTree/Strict.hs

```diff
@@ -0,0 +1,250 @@
+-- is 'False' for the split immediately before @x@ and 'True' for the
+-- split just after it:
+--
+-- <<images/search.svg>>
+--
+-- In this situation, @'search' p t@ returns such an element @x@ and the
+-- pieces @l@ and @r@ of the sequence to its left and right respectively.
+-- That is, it returns @'Position' l x r@ such that
+--
+-- * @l >< (x <| r) = t@
+--
+-- * @not (p (measure l) (measure (x <| r))@
+--
+-- * @p (measure (l |> x)) (measure r)@
+--
+-- For predictable results, one should ensure that there is only one such
+-- point, i.e. that the predicate is /monotonic/ on @t@.
+--
+-- @since 0.1.2.0
+search :: (Measured v a)
+       => (v -> v -> Bool) -> StrictFingerTree v a -> SearchResult v a
+search p (SFT ft) = fromOriginalSearchResult (FT.search p ft)
+
+{-------------------------------------------------------------------------------
+  Splitting
+-------------------------------------------------------------------------------}
+
+-- | /O(log(min(i,n-i)))/. Split a sequence at a point where the predicate
+-- on the accumulated measure of the prefix changes from 'False' to 'True'.
+--
+-- For predictable results, one should ensure that there is only one such
+-- point, i.e. that the predicate is /monotonic/.
+split :: (Measured v a)
+      => (v -> Bool)
+      -> StrictFingerTree v a
+      -> (StrictFingerTree v a, StrictFingerTree v a)
+split p (SFT ft) = (SFT left, SFT right)
+  where
+    (left, right) = FT.split p ft
+
+-- | /O(log(min(i,n-i)))/.
+-- Given a monotonic predicate @p@, @'takeUntil' p t@ is the largest
+-- prefix of @t@ whose measure does not satisfy @p@.
+--
+-- *  @'takeUntil' p t = 'fst' ('split' p t)@
+takeUntil :: (Measured v a)
+          => (v -> Bool) -> StrictFingerTree v a -> StrictFingerTree v a
+takeUntil p (SFT ft) = SFT (FT.takeUntil p ft)
+
+-- | /O(log(min(i,n-i)))/.
+-- Given a monotonic predicate @p@, @'dropUntil' p t@ is the rest of @t@
+-- after removing the largest prefix whose measure does not satisfy @p@.
+--
+-- * @'dropUntil' p t = 'snd' ('split' p t)@
+dropUntil :: (Measured v a)
+          => (v -> Bool) -> StrictFingerTree v a -> StrictFingerTree v a
+dropUntil p (SFT ft) = SFT (FT.dropUntil p ft)
+
+{-------------------------------------------------------------------------------
+  Transformation
+-------------------------------------------------------------------------------}
+
+-- | /O(n)/. The reverse of a sequence.
+reverse :: (Measured v a) => StrictFingerTree v a -> StrictFingerTree v a
+reverse (SFT ft) = SFT (FT.reverse ft)
+
+{-------------------------------------------------------------------------------
+  Maps
+-------------------------------------------------------------------------------}
+
+-- | Like 'fmap', but with constraints on the element types.
+fmap' :: (Measured v1 a1, Measured v2 a2)
+      => (a1 -> a2) -> StrictFingerTree v1 a1 -> StrictFingerTree v2 a2
+fmap' f (SFT ft) = SFT (FT.fmap' f ft)
```

Comment:
I don't think this is strict: each element `x` in the finger tree will be a thunk `f x`.

Reply:
Yes, I believe you're right. Currently re-thinking how I should go about this.

Reply:
I think this is tricky, normally one would get strictness from a bang in the data constructor, which we don't do here.  Adding a bang to result of `f` will not be right (evaluate `f x` when you evaluate `f x`; by strictness `x` is already evaluated). You probably need to re-implement `fmap` without using `FingerTree` `Functor` instance.

Reply:
So I went through this with @dcoutts yesterday and he suggested a good solution, at the cost of some performance. The solution is similar to one that @mrBliss suggested on https://github.com/input-output-hk/cardano-ledger/pull/637#discussion_r326649968.

However, the problem with @mrBliss' solution is that the `mappend` implementation for `()` is completely lazy: `_ <> _ = ()`. I didn't know this at the time so I went with another solution.

In this case, I went with this:

```
-- | Like 'fmap', but with constraints on the element types.
fmap' :: (Measured v1 a1, Measured v2 a2)
      => (a1 -> a2) -> StrictFingerTree v1 a1 -> StrictFingerTree v2 a2
fmap' f (SFT ft) = foldMap (\x -> seq x StrictUnit) sft `seq` sft
  where
    sft = SFT (FT.fmap' f ft)

data StrictUnit = StrictUnit

instance Semigroup StrictUnit where
  StrictUnit <> StrictUnit = StrictUnit

instance Monoid StrictUnit where
  mempty = StrictUnit
```